### DISCUSSION TITLE
Timer callback under heavy load

### DISCUSSION THREAD

**Discussion by pulkin** [micropython/micropython#10731] [ESP32]
I have the following `pulse` method for haptic feedback:
```python
    def pulse(self, time, blocking=False):
        vibro.on()
        if blocking:
            sleep(time)
            vibro.off()
        else:
            def _callback(timer, vibro=self.vibro):
                vibro.off()
            self.timer.init(period=int(time * 1000), mode=machine.Timer.ONE_SHOT, callback=_callback)
```
What I find is that for smaller `time=0.01` the non-blocking `vibro.off()` is fired inconsistently. It feels like under heavy load the callback will be delayed. For example, the following piece of code will cause a strong feedback on first run (when the module is compiled) and a "normal" feedback on subsequent runs.
```python
vibro.pulse(0.01)
import heavy_python_module
```
Any idea how to fix it?

**Comment by robert-hh** [comment]
machine.Timer schedules the callback functions, in contrast to a hard interrupt. That has fewer restrictions for the callback code, but the scheduling can be delayed in certain situations. It seems ghat during code import the scheduler is not called.
